src/slot_provider/clock.rs

```diff
@@ -1,88 +1,185 @@
 //! Clock based SlotProvider
 //! When a fn does not finish within a slot (12 seconds), the next call is made concurrently, up to a buffer limit. When the buffer limit is reached, the passed fn will not be called for a new slot, but wait in a FIFO queue until a previous slot has finished.
-use ethers::providers::StreamExt;
-use eyre::Result;
-use futures::Future;
-use tokio::time::{interval, timeout, Duration};
-
-use crate::slot_provider::{
-    wait_until_slot_start, Slot, SlotProvider, GENESIS_SLOT_TIME, SLOT_PERIOD_SECONDS,
-};
+use async_trait::async_trait;
+use chrono::Utc;
+use futures::stream::StreamExt;
+use std::collections::VecDeque;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use tokio::sync::{mpsc, Mutex};
+use tokio::time::{interval, Duration};
 use tokio_stream::wrappers::IntervalStream;
 
+use super::{Slot, SlotProvider};
+
+const SLOT_PERIOD_SECONDS: u32 = 12;
+const MAX_SLOTS: usize = 32;
+
+/// Waits until the start of the next slot.
+/// Much of our code depends on what the current slot is, and wants to answer as fast as possible,
+/// therefore we sometimes want to align our code with the start of the slot.
+async fn wait_until_next_slot() {
+    let current_slot = Slot::from_timestamp(Utc::now().timestamp() as u64).unwrap();
+    let next_slot = Slot {
+        number: current_slot.number + 1,
+    };
+    let next_slot_start = next_slot.to_date_time();
+    // NOTE: doesn't account for leap seconds.
+    let seconds_until_next_slot = next_slot_start.timestamp() - Utc::now().timestamp();
+
+    tokio::time::sleep(tokio::time::Duration::from_secs(
+        seconds_until_next_slot as u64,
+    ))
+    .await;
+}
+
 pub struct SystemClockSlotProvider {
-    num_slots: Option<usize>,
+    counter: AtomicUsize,
```

Comment:
Using a counter here. Feels more rusty, but don't really care. Could be stream with .take() all the same I think. A nice benefit from that would be no `break`.

Reply:
I like to avoid `break`. Break is a jump statement. I read that back in the day, assembly era, there was `goto`. Super powerful, and great at making it incredibly hard to reason about control flow. It's a bit hard to explain. Has to do with how complex the decision tree is you have to keep in your head to be able to reason about a piece of code. Break can make things quite complex depending on the lines that come after it.

You're version with `.take` didn't have that complexity.

I asked AI for good measure. Nice to see point 1 is what I was getting at. 2. is kinda the same as one, and 3. and 4. don't concern me, but I agree with 1.!

<img width="850" alt="Screenshot 2023-07-06 at 18 08 10" src="https://github.com/ultrasoundmoney/oracle-client/assets/2011351/9393a170-2b03-40e7-b527-f28f99a651b5">

Reply:
Got it, good to know. As far as I can see there is no more "break" in the latest version.